File: server/src/main/java/com/example/demo/model/Room.java
package com.example.demo.model;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import lombok.Getter;
import lombok.Setter;
import lombok.ToString;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.Instant;
import java.util.HashSet;
import java.util.Set;
import java.util.UUID;


@Entity
@EntityListeners(AuditingEntityListener.class)
@Table(name = "rooms")
@Getter
@Setter
@ToString
public class Room {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(updatable = false, nullable = false)
    private UUID id;

    @Column
    private String name;

    @NotNull
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private RoomType type = RoomType.DIRECT;

    @NotNull
    @Column(nullable = false, updatable = false)
    private UUID createdBy;

    @CreatedDate
    @Column(nullable = false, updatable = false)
    private Instant createdAt;

    @LastModifiedDate
    @Column(nullable = false)
    private Instant updatedAt;

    @OneToMany(mappedBy = "room", cascade = CascadeType.ALL, orphanRemoval = true)
    private Set<RoomParticipant> participants = new HashSet<>();
}

File: client/src/types/room.ts
export type RoomType = "DIRECT" | "GROUP";

export interface User {
  id: string;
  name: string;
  email?: string;
}

export interface Room {
  id: string;
  name: string | null;
  type: RoomType;
  createdAt: string;
  createdBy: string;
  updatedAt: string;
  participants: User[];
}

export function isRoom(obj: any): obj is Room {
  return (
    obj !== null &&
    typeof obj === "object" &&
    typeof obj.id === "string" &&
    (typeof obj.name === "string" || obj.name === null) &&
    (obj.type === "DIRECT" || obj.type === "GROUP") &&
    typeof obj.createdAt === "string" &&
    typeof obj.updatedAt === "string" &&
    typeof obj.createdBy === "string" &&
    Array.isArray(obj.participants) &&
    obj.participants.every(
      (p: any) => p && typeof p === "object" && typeof p.id === "string" && typeof p.name === "string"
    )
  );
}
